### ISSUE TITLE
`TypeAliasForwardRef` nested in `typing` constructs

### ISSUE THREAD

**Issue by dalcinl** [sphinx-doc/sphinx#10333]
### Describe the bug

I have problems rendering function signatures when type aliases are used within other `typing` constructs, let say `Union[MyAlias1, MyAlias2]`.

I had to monkeypatch `TypeAliasForwardRef` the following way:

```python
from sphinx.util import inspect
inspect.TypeAliasForwardRef.__repr__ = lambda self: self.name
inspect.TypeAliasForwardRef.__hash__ = lambda self: hash(self.name)
```

Without it, two different things happen:

* The type `TypeAliasForwardRef` "escapes" from stringifycation and produces undefined reference warnings. So I had to add `__repr__`.
* Without `__hash__`, `typing.get_type_hints()` raises, and then `sphinx.util.inspect.signature()` returns am empty `Signature` object, which in turns affects further processing and produces suboptimal rendering (e.g., `Callable`, `Union`, etc. text without links to Python's docs).

### How to Reproduce

Clone mpi4py repo:
```
$ git clone https://github.com/mpi4py/mpi4py
$ cd mpi4py
```

Apply the following patch (to remove my monkeypatching):
```diff
diff --git a/docs/source/conf.py b/docs/source/conf.py
index a9d20362..8fd56e07 100644
--- a/docs/source/conf.py
+++ b/docs/source/conf.py
@@ -159,8 +159,8 @@ def _patch_domain_python():
     make_xref_orig = python.PyXrefMixin.make_xref
     python.PyXrefMixin.make_xref = make_xref
 
-    inspect.TypeAliasForwardRef.__repr__ = lambda self: self.name
-    inspect.TypeAliasForwardRef.__hash__ = lambda self: hash(self.name)
+    #inspect.TypeAliasForwardRef.__repr__ = lambda self: self.name
+    #inspect.TypeAliasForwardRef.__hash__ = lambda self: hash(self.name)
```

Finally run:
```
tox -e docs
```

### Your project

https://github.com/mpi4py/mpi4py

### OS

Linux

### Python version

3.10.2

### Sphinx version

4.5.0

### Sphinx extensions

sphinx.ext.autodoc, sphinx.ext.autosummary, sphinx.ext.napoleon, sphinx.ext.intersphinx

**Comment by orenbenkiki** [comment]
Just run into this problem. Thanks for reporting it!

**Comment by dalcinl** [comment]
@orenbenkiki Fixed in #10353.

**Comment by orenbenkiki** [comment]
Thanks!
